Translation of Python UDFs doc (#861)
ytsaurus/ytsaurus: yt/docs/en/_includes/user-guide/query-tracker/system-python-udf.md

# Python user-defined functions (UDFs) in YQL

## Introduction

Sometimes it can be difficult to describe data processing tasks using only a declarative language like SQL. You may find it more convenient to use an imperative language as well. For this purpose, YQL supports the creation and use of user-defined functions (UDFs).

## Python UDF

With the Python UDF feature, you can define and use Python functions inside a query.

## How to use

To use a Python UDF in a YQL query, follow these steps:

1. Define your function in a Python script and include this script as a string in your query. This function will be called as a UDF.
2. In your query, declare the function name and its signature (input and output data types).
3. Call the function in the query body and pass the necessary arguments to it.

Each of these steps is explained in detail below.

### Write Python script

You can add your script as:

- Regular string literal. For example, `$script = "def foo(): return 'bar'"`.

> - Multi-line string literal (a YQL extension). You must enclose this literal in the `@@` characters, which are similar to triple quotes in Python. The `#py` comment enables Python syntax highlighting in the YQL editor:

Reviewer: ```suggestion
- Multi-line string literal (YQL extension). You must enclose this literal in the `@@` characters, which are similar to triple quotes in Python. The `#py` comment enables Python syntax highlighting in the YQL editor:
```

An alternative would be to use `an`

  ```yql
  $script = @@#py
  def foo():
      return b'bar'
  @@
  ```

> - Named file attached to the query. You must include the file content in your query by using the [FileContent](../../../yql/builtins/basic.md#file-content-path) function. For example, `$script = FileContent("foo.py")`.

Reviewer: ```suggestion
- Named file attached to the query. You can include the file content in your query by using the [FileContent](../../../yql/builtins/basic.md#file-content-path) function. For example, `$script = FileContent("foo.py")`.
```

### Declare function name and signature

Since Python is a dynamically typed language and YQL is statically typed, you need to define the types in advance. The name and signature of the function called from the Python script must be known before executing a YQL query. If the types you declare in the function signature do not match the actual data that the function returns, you get a runtime error.

Example of declaring a function name and signature:
`$f = SystemPython3_8::foo(Callable<()->String>, $script);`. In this example:

- `$f` and `$script`: [Named YQL expressions](../../../yql/syntax/expressions.md#named-nodes).
  - `$f`: Ready-to-use function.
  - `$script`: Python script mentioned in the previous section.

> - `SystemPython3_8::`: Fixed prefix for declaring a Python function. `System` means that Python must be used from the environment at runtime. `3_8` indicates the Python version to be used.

Reviewer: ```suggestion
- `SystemPython3_8::`: Fixed prefix for declaring a Python function. `System` means that Python will be taken from the environment at runtime. `3_8` indicates the Python version to be used.
```

| **Name** | **Signature declaration** | **Signature example** | **Representation in Python** |
| --- | --- | --- | --- |
| List | `List<Type>` | `List<Int32>` | `list-like object` [(learn more)](#python-container-wrappers) |
| Dictionary | `Dict<KeyType,ValueType>` | `Dict<String,Int32>` | `dict-like object` [(learn more)](#python-container-wrappers) |
| Tuple | `Tuple<Type1,...,TypeN>` | `Tuple<Int32,Int32>` | `tuple` |
| Structure | `Struct<Name1:Type1,...,NameN:TypeN>` | `Struct<Name:String,Age:Int32>` | `StructSequence` |
| Stream | `Stream<Type>` | `Stream<Int32>` | `generator` |
| Variant on tuple | `Variant<Type1,Type2>` | `Variant<Int32,String>` | `tuple with index and object` |
| Variant on structure | `Variant<Name1:Type1,Name2:Type2>` | `Variant<value:Int32,error:String>` | `tuple with field name and object` |

You can nest containers into each other. For example, `List<Tuple<Int32,Int32>>`.

### Specifics of objects for lists and dictionaries passed to functions {#python-container-wrappers}

When you use `List` and `Dict`, the read-only `yql.TList` and `yql.TDict` objects are passed to the function arguments.

> The objects' specifics:

Reviewer: ```suggestion
The object's specifics:
```

> - They are read-only. This means that you can't make any changes to these objects without copying them because another part of the current query, for example, a neighboring function, may use the same object.

Reviewer: ```suggestion
- They are read-only. This means that you can't make any changes to these objects without copying them because another part of the current query may use the same object. For example, a neighboring function.
```

$script = @@#py
import sys
import tensorflow

def foo():
    return "Python version: " + sys.version + ".\nTensorflow version: " + tensorflow.__version__
@@;

$udf = SystemPython3_11::foo(Callable<()->String>, $script);
SELECT $udf();
-- Python version: 3.11.0rc1 (main, Aug 12 2022, 10:02:14) [GCC 11.2.0].
-- Tensorflow version: 2.16.1
```

## Specifics

When using UDFs, the standard output (stdout) is used for system purposes. Therefore, you must not use it.

## Examples

### Hello World

```yql
$script = @@#py
def hello(name):
    return b'Hello, %s!' % name
@@;

$udf = SystemPython3_8::hello(Callable<(String)->String>, $script);

SELECT $udf("world"); -- "Hello, world!"
```

### String and number concatenation

```yql
$script = @@#py
def concat(a, b):
    return a.decode("utf-8") + str(b)
@@;
$concat = SystemPython3_8::concat(Callable<(String?, Int64?)->String>, $script);

SELECT $concat(name, age) FROM `//tmp/sample`;
```

### Retrieving the latest list of `yql.TList` methods {#tlist-methods}

```yql
$u = SystemPython3_8::get_dir(Callable<(List<Int32>)->List<String>>, @@#py
def get_dir(lst):
  return dir(lst)
@@);
SELECT $u(AsList());
```

> ### Calling Llama

Reviewer: ```suggestion
### Inferencing Llama
```
